File: modules/philoi-focus-nudge/ios/PhiloiFocusNudgeModule.swift
// The app-side half of Focus Nudge (APP_BLOCKER_SPEC §A/§B/§D, FOCUS_NUDGE_SETUP.md Part B).
//
// Four jobs, and only four:
//   1. ask for Screen Time authorization,
//   2. put Apple's FamilyActivityPicker up so the user chooses what to guard,
//   3. hand Cindy's pre-fetched line to the App Group so the shield can draw it offline,
//   4. arm on lock-in start and disarm on lock-in end.
//
// The shield itself lives in three extensions (targets/shield-configuration, targets/shield-action,
// targets/device-activity-monitor) because iOS renders it out of process — the same reason the
// Live Activity's UI is its own target. Everything those three read comes from here, through the
// App Group, because they cannot network. FocusNudgeShared.swift is the contract.
//
// 🔒 THE TOKENS ARE OPAQUE AND STAY THAT WAY. A FamilyActivitySelection is a set of tokens Apple
// deliberately makes unresolvable to app identities, and nothing below tries. The only thing this
// file ever reports upward is how MANY were picked.

import DeviceActivity
import ExpoModulesCore
import FamilyControls
import ManagedSettings
import SwiftUI

/// Mirrors ArmOptions in modules/philoi-focus-nudge/index.ts. Field names must match the JS keys
/// exactly — Expo decodes Records by name, and a typo reads as a silent default.
struct FocusNudgeArmOptions: Record {
  /// The failsafe ceiling (§D). The DeviceActivity schedule ends here no matter what, so a
  /// force-quit mid-session cannot leave someone shielded indefinitely.
  @Field var maxMinutes: Double = 720
  /// How long a "continue anyway" holds the shield down before the monitor's usage-threshold event
  /// puts it back. Kept in step with the payload's own `deferMs`.
  @Field var deferMinutes: Double = 10
}

public class PhiloiFocusNudgeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PhiloiFocusNudge")

    // ─────────────────────────── authorization ───────────────────────────

    /// "notDetermined" | "denied" | "approved". Read fresh every time: the user can revoke Screen
    /// Time access in Settings while the app is running, and §"Edge cases" requires us to notice
    /// rather than crash.
    Function("authorizationStatus") { () -> String in
      switch AuthorizationCenter.shared.authorizationStatus {
      case .approved: return "approved"
      case .denied: return "denied"
      case .notDetermined: return "notDetermined"
      @unknown default: return "notDetermined"
      }
    }

    /// Resolves with the status AFTER the prompt. Denial is not an error — it is a normal answer
    /// that turns the feature off, and it must never block anyone from locking in.
    AsyncFunction("requestAuthorization") { () -> String in
      do {
        try await AuthorizationCenter.shared.requestAuthorization(for: .individual)
        return "approved"
      } catch {
        // Thrown on refusal as well as on genuine failure, and the two are indistinguishable here.
        // Report the status the system actually holds rather than guessing from the error.
        return AuthorizationCenter.shared.authorizationStatus == .approved ? "approved" : "denied"
      }
    }

    // ─────────────────────────── the picker ───────────────────────────

    /// Apple's own FamilyActivityPicker, presented as a sheet.
    ///
    /// A sheet rather than an embedded RN view component: FamilyActivityPicker is SwiftUI, its
    /// contents are privileged, and hosting it inside the RN view tree buys nothing — the user
    /// picks apps once, in a modal, from the Focus Nudge settings screen.
    ///
    /// Resolves with the counts (never identities). Cancelling resolves with the counts UNCHANGED
    /// rather than rejecting, because "I looked and changed nothing" is not an error.
    AsyncFunction("presentPicker") { (promise: Promise) in
      DispatchQueue.main.async {
        guard let presenter = self.appContext?.utilities?.currentViewController() else {
          promise.reject("no_view_controller", "No view controller to present the picker from.")
          return
        }

        let host = FocusNudgePickerHost(
          initial: FocusNudgeSelection.load() ?? FamilyActivitySelection()
        ) { result in
          presenter.presentedViewController?.dismiss(animated: true)
          if let saved = result {
            FocusNudgeSelection.save(saved)
            // Re-apply immediately when a session is already running, so changing the picked apps
            // mid-lock-in takes effect now rather than at the next session.
            if FocusNudgeState.armedAtMs > 0 { FocusNudgeShield.arm() }
          }
          promise.resolve(Self.counts())
        }

        let controller = UIHostingController(rootView: host)
        controller.modalPresentationStyle = .formSheet
        presenter.present(controller, animated: true)
      }
    }

    /// { applications, categories, webDomains } — counts only. This is the whole of what the app
    /// is ever allowed to know about the selection.
    Function("selectionCounts") { () -> [String: Int] in
      Self.counts()
    }

    Function("clearSelection") {
      FocusNudgeSelection.clear()
      FocusNudgeShield.disarm()
    }

    // ─────────────────────────── the handoff ───────────────────────────

    /// Cache the nudge copy for the shield to read.
    ///
    /// Takes a JSON string rather than a Record, on purpose: this payload is authored end-to-end in
    /// JS (src/lib/focus-nudge.ts) so the wording ships over OTA, and the Swift on both sides of
    /// the App Group treats it as opaque text. Nothing here validates it — FocusNudgePayload.load()
    /// in the shield falls back to its own built-in card if it cannot be parsed, which is the
    /// behaviour we want anyway for a partially-written or future-shaped payload.
    Function("writePayload") { (json: String) in
      focusNudgeDefaults?.set(json, forKey: FocusNudgeKey.payload)
    }

    /// How many times the shield has been presented inside the window — written by the
    /// ShieldConfiguration extension, read back here so the next coach call knows it is dealing
    /// with repeated retreat (§C-safety graduated response).
    Function("retreatCount") { (windowMs: Double) -> Int in
      FocusNudgeState.retreats(window: windowMs).count
    }

    // ─────────────────────────── arm / disarm ───────────────────────────

    Function("isArmed") { () -> Bool in
      FocusNudgeState.armedAtMs > 0
    }

    /// Arm for the current lock-in.
    ///
    /// The shield goes up HERE, immediately, rather than waiting for the monitor extension —
    /// intervalDidStart is delivered on the system's schedule and an app opened three seconds after
    /// starting a session must already be covered. The DeviceActivity schedule registered alongside
    /// it exists for the two things this process cannot do: disarm after a force-quit (§D's
    /// failsafe) and re-arm once a "continue anyway" cooldown has been used up.
    AsyncFunction("arm") { (options: FocusNudgeArmOptions) -> Bool in
      guard AuthorizationCenter.shared.authorizationStatus == .approved else { return false }
      guard let selection = FocusNudgeSelection.load(),
        !(selection.applicationTokens.isEmpty && selection.categoryTokens.isEmpty
          && selection.webDomainTokens.isEmpty)
      else { return false }

      // A new session starts clean: a previous session's deferral must not silently swallow the
      // first nudge of this one, and its retreat history must not escalate this one early.
      FocusNudgeState.deferredUntilMs = 0
      FocusNudgeState.clearRetreats()
      FocusNudgeState.armedAtMs = focusNudgeNowMs()
      FocusNudgeShield.arm(force: true)

      let center = DeviceActivityCenter()
      center.stopMonitoring([.focusNudgeSession])

      let calendar = Calendar.current
      let now = Date()
      // A MINUTE IN THE PAST, deliberately. A DeviceActivitySchedule is expressed as times of day,
      // so a start set to exactly "now" is ambiguous about whether this instant is inside the
      // window or a hair before it — and if the system resolves it as "starts later", the failsafe
      // interval (and its intervalDidEnd) slides a whole day out. Backdating the start removes the
      // question: we are unambiguously inside the window the moment monitoring begins.
      let start = now.addingTimeInterval(-60)
      let schedule = DeviceActivitySchedule(
        intervalStart: calendar.dateComponents([.hour, .minute, .second], from: start),
        intervalEnd: calendar.dateComponents(
          [.hour, .minute, .second], from: start.addingTimeInterval(options.maxMinutes * 60)),
        // Never repeats. This window belongs to ONE lock-in; a repeating schedule would shield the
        // same hours tomorrow whether or not anyone was working.
        repeats: false
      )

      // Usage threshold, not wall clock: this fires once the user has actually spent deferMinutes
      // inside the guarded apps, which is the moment a "I really need a sec" has plainly turned
      // into a scroll. The monitor re-arms then.
      let deferLapsed = DeviceActivityEvent(
        applications: selection.applicationTokens,
        categories: selection.categoryTokens,
        webDomains: selection.webDomainTokens,
        threshold: DateComponents(minute: Int(options.deferMinutes))
      )

      do {
        try center.startMonitoring(
          .focusNudgeSession, during: schedule, events: [.focusNudgeDeferLapsed: deferLapsed])
      } catch {
        // The shield is already up — monitoring is the failsafe, not the mechanism. Losing it
        // costs us the force-quit sweep and the re-arm, not the feature, so this is reported and
        // swallowed rather than failing the arm and leaving the user unguarded.
        NSLog("[PhiloiFocusNudge] startMonitoring failed: \(error.localizedDescription)")
      }
      return true
    }

    /// End of session. Unconditional and idempotent — see the sweep note in the monitor extension.
    AsyncFunction("disarm") { () -> Void in
      FocusNudgeShield.disarm()
      FocusNudgeState.armedAtMs = 0
      FocusNudgeState.deferredUntilMs = 0
      FocusNudgeState.clearRetreats()
      DeviceActivityCenter().stopMonitoring([.focusNudgeSession])
    }

    /// Put the shield back if a cooldown has lapsed while the app was away. Cheap enough to call on
    /// every foreground; a no-op unless something has actually changed.
    AsyncFunction("reconcile") { () -> Bool in
      guard FocusNudgeState.armedAtMs > 0 else { return false }
      guard AuthorizationCenter.shared.authorizationStatus == .approved else {
        // Permission revoked mid-session (§"Edge cases"). Leave nothing behind and let JS surface
        // the "Focus Nudge is off" state.
        FocusNudgeShield.disarm()
        FocusNudgeState.armedAtMs = 0
        return false
      }
      if FocusNudgeState.isDeferred { return false }
      FocusNudgeShield.arm()
      return true
    }
  }

  private static func counts() -> [String: Int] {
    guard let selection = FocusNudgeSelection.load() else {
      return ["applications": 0, "categories": 0, "webDomains": 0]
    }
    return [
      "applications": selection.applicationTokens.count,
      "categories": selection.categoryTokens.count,
      "webDomains": selection.webDomainTokens.count,
    ]
  }
}

/// The SwiftUI wrapper around Apple's picker. Its own type so the selection binding lives
/// somewhere with a lifetime — a FamilyActivityPicker bound to a local var in a closure loses
/// every tap.
private struct FocusNudgePickerHost: View {
  @State private var selection: FamilyActivitySelection
  /// nil = cancelled, and the stored selection is left exactly as it was.
  private let onFinish: (FamilyActivitySelection?) -> Void

  init(initial: FamilyActivitySelection, onFinish: @escaping (FamilyActivitySelection?) -> Void) {
    _selection = State(initialValue: initial)
    self.onFinish = onFinish
  }

  var body: some View {
    NavigationView {
      FamilyActivityPicker(selection: $selection)
        .navigationTitle("Apps to nudge on")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
          ToolbarItem(placement: .cancellationAction) {
            Button("Cancel") { onFinish(nil) }
          }
          ToolbarItem(placement: .confirmationAction) {
            Button("Done") { onFinish(selection) }
          }
        }
    }
  }
}

File: modules/philoi-focus-nudge/android/src/main/java/expo/modules/philoifocusnudge/PhiloiFocusNudgeModule.kt
package expo.modules.philoifocusnudge

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import android.text.TextUtils
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

// The app-side half of Focus Nudge on Android — the counterpart of PhiloiFocusNudgeModule.swift,
// and deliberately the same shape so src/lib/focus-nudge.ts can talk to either one.
//
// Its jobs, in the same order as iOS:
//   1. report whether the two permissions are granted, and open the right Settings screen,
//   2. let the user pick from the curated list of guardable apps,
//   3. hand Cindy's pre-fetched line to SharedPreferences so the overlay can draw it offline,
//   4. arm on lock-in start and disarm on lock-in end.
//
// WHERE THE TWO PLATFORMS GENUINELY DIVERGE, and it is worth naming because it is the whole design:
//
//   · PERMISSIONS. iOS has one prompt this module can raise itself. Android has two, and NEITHER
//     can be granted from inside an app — an AccessibilityService cannot be enabled
//     programmatically by any means (that restriction is the point of the permission), and
//     SYSTEM_ALERT_WINDOW is a Settings trip too. So there is no requestAuthorization() here. There
//     are two "open the right Settings page" intents and a screen that explains why, which is all
//     the platform allows.
//
//   · ARMING. iOS applies a ManagedSettingsStore and registers a DeviceActivity schedule, and needs
//     a monitor extension to sweep up after a force-quit. Android arms by writing a timestamp: the
//     accessibility service reads it on every window change, so "armed" is a fact on disk rather
//     than a state applied to the system. Nothing to leak, nothing to sweep — a force-quit leaves a
//     stale `armedAtMs`, and the very next launch's cold-start disarm in FocusNudgeSync clears it.
//     That is why there is no Android equivalent of the 12-hour failsafe window.
//
//   · WHAT WE KNOW. Apple hands back opaque tokens; Android hands back package names. Both sides
//     report only the COUNT upward, so the analytics surface is identical (see FocusNudgeKey.GUARDED).

/** Mirrors FocusNudgeArmOptions in modules/philoi-focus-nudge/index.ts and its Swift twin. */
class FocusNudgeArmOptions : Record {
  /**
   * The iOS failsafe ceiling. Accepted and ignored here — see "ARMING" above; Android has no
   * applied system state that could outlive the process. Kept in the record rather than dropped so
   * one JS call site serves both platforms.
   */
  @Field var maxMinutes: Double = 720.0

  /** What a "continue anyway" buys. This is the one the Android guard actually enforces. */
  @Field var deferMinutes: Double = 10.0
}

class PhiloiFocusNudgeModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("No react context")

  override fun definition() = ModuleDefinition {
    Name("PhiloiFocusNudge")

    // ─────────────────────────── permissions ───────────────────────────

    /**
     * The shared three-state contract, collapsed onto Android's two switches.
     *
     * "denied" is never returned. On iOS it is a real, distinguishable answer — the user was asked
     * and said no. Android never asks: both permissions are toggles in Settings that have simply
     * not been turned on yet, and there is no way to tell "hasn't got round to it" from "went and
     * turned it off". Reporting "denied" would make the setup screen say "turned off in Settings"
     * to someone who has never seen a prompt.
     */
    Function("authorizationStatus") {
      if (hasAccessibility(context) && Settings.canDrawOverlays(context)) "approved"
      else "notDetermined"
    }

    /** The two switches, separately — the setup screen walks the user through them one at a time. */
    Function("permissions") {
      mapOf(
        "accessibility" to hasAccessibility(context),
        "overlay" to Settings.canDrawOverlays(context),
      )
    }

    /**
     * The system's accessibility list. There is no way to deep-link reliably to Philoi's own row:
     * the `:settings:fragment_args_key` extra that does it is undocumented and OEM skins move or
     * ignore it, so a wrong guess lands on a blank screen. The list plus the on-screen instruction
     * beats a deep link that works on Pixel and nowhere else.
     */
    Function("openAccessibilitySettings") {
      launchSettings(Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS))
    }

    Function("openOverlaySettings") {
      launchSettings(
        Intent(
          Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
          Uri.parse("package:" + context.packageName),
        )
      )
    }

    // ─────────────────────────── the picker ───────────────────────────

    /**
     * Which of the candidate packages are actually on this phone.
     *
     * The candidates come from JS, out of modules/philoi-focus-nudge/android-guarded-apps.json —
     * the same file the config plugin turns into the manifest's <queries> allow-list. That is the
     * whole reason this app needs no QUERY_ALL_PACKAGES: it never enumerates anything, it asks
     * about a fixed list of ~14 names it declared up front. Anything outside that list is invisible
     * to this call, which is exactly the property the allow-list is meant to have.
     */
    Function("installedPackages") { candidates: List<String> ->
      val packageManager = context.packageManager
      candidates.filter { candidate ->
        try {
          packageManager.getPackageInfo(candidate, 0)
          true
        } catch (e: Exception) {
          // Not installed, or not in <queries>. Indistinguishable from here and the same outcome:
          // the app is not offered.
          false
        }
      }
    }

    /**
     * Shaped like the iOS selection counts so focusNudgeSelectionSize() works unchanged on both.
     * Android has no notion of guarding a category or a web domain, so those are structurally zero.
     */
    Function("selectionCounts") {
      mapOf(
        "applications" to FocusNudgeState.guardedPackages(context).size,
        "categories" to 0,
        "webDomains" to 0,
      )
    }

    /**
     * The current selection, by package name.
     *
     * Yes, this hands package names up to JS — unavoidable, because JS draws the picker and a
     * picker has to know what is ticked. The discipline the iOS side gets for free from Apple's
     * opaque tokens is kept by hand here instead: src/app/focus-nudge.tsx tracks these locally and
     * reports only `count` to analytics, exactly as the iOS screen does. Nothing named ever leaves
     * the phone.
     */
    Function("guardedPackages") { FocusNudgeState.guardedPackages(context).toList() }

    Function("setGuardedPackages") { packages: List<String> ->
      FocusNudgeState.setGuardedPackages(context, packages)
      // Changing the list mid-session takes effect on the next window change, with no re-arm: the
      // service reads the set fresh on every event. An app removed from the list while its nudge is
      // on screen is the one case that needs a hand, so drop the overlay too.
      if (packages.isEmpty()) FocusNudgeOverlay.hide(context)
    }

    Function("clearSelection") {
      FocusNudgeState.setGuardedPackages(context, emptyList())
      FocusNudgeOverlay.hide(context)
    }

    // ─────────────────────────── the handoff ───────────────────────────

    /**
     * Cache the nudge copy for the overlay to read.
     *
     * A JSON string rather than a Record, for the same reason as iOS: the payload is authored
     * end-to-end in JS (src/lib/focus-nudge.ts) so the wording ships over OTA, and both native
     * sides treat it as opaque text. Nothing validates it here — FocusNudgePayload.load() falls
     * back to its built-in card on anything it cannot parse, which is the behaviour we want for a
     * half-written or future-shaped payload anyway.
     */
    Function("writePayload") { json: String ->
      focusNudgePrefs(context).edit().putString(FocusNudgeKey.PAYLOAD, json).apply()
    }

    /** Presentations inside the window — written by the overlay, read here for the next coach call. */
    Function("retreatCount") { windowMs: Double ->
      FocusNudgeState.retreats(context, windowMs).size
    }

    // ─────────────────────────── arm / disarm ───────────────────────────

    Function("isArmed") { FocusNudgeState.isArmed(context) }

    /**
     * Arm for the current lock-in.
     *
     * False is the ordinary answer, not an error: no accessibility grant, no overlay grant, or
     * nothing picked. All three mean "the feature is off", and none of them may ever stop someone
     * locking in.
     */
    AsyncFunction("arm") { options: FocusNudgeArmOptions ->
      if (!hasAccessibility(context)) return@AsyncFunction false
      if (!Settings.canDrawOverlays(context)) return@AsyncFunction false
      if (FocusNudgeState.guardedPackages(context).isEmpty()) return@AsyncFunction false
      // A new session starts clean: a previous session's deferral must not silently swallow the
      // first nudge of this one, and its retreat history must not escalate this one early.
      FocusNudgeState.arm(context, (options.deferMinutes * 60_000).toLong())
      true
    }

    /** End of session. Unconditional and idempotent. */
    AsyncFunction("disarm") {
      FocusNudgeState.disarm(context)
      // A nudge left on screen after its session ended is the only genuinely harmful failure this
      // feature has. Cheap to prevent, so it is prevented unconditionally rather than only on the
      // paths we think can reach here with the overlay up.
      FocusNudgeOverlay.hide(context)
    }

    /**
     * Called on every foreground. On iOS this re-applies a shield a cooldown took down; here there
     * is no applied state to restore — the deferral expires on its own clock — so the only work is
     * noticing that a permission was revoked mid-session and standing down honestly rather than
     * leaving the lock-in screen claiming a guard that is no longer there.
     */
    AsyncFunction("reconcile") {
      if (!FocusNudgeState.isArmed(context)) return@AsyncFunction false
      if (!hasAccessibility(context) || !Settings.canDrawOverlays(context)) {
        FocusNudgeState.disarm(context)
        FocusNudgeOverlay.hide(context)
        return@AsyncFunction false
      }
      !FocusNudgeState.isDeferred(context)
    }
  }

  // ─────────────────────────── plumbing ───────────────────────────

  /**
   * Whether OUR service is in the system's enabled list.
   *
   * Settings.Secure rather than AccessibilityManager.getEnabledAccessibilityServiceList(), because
   * the latter answers "is any service with this feedback type running" and we need "is this exact
   * component enabled" — the difference matters on a phone that has TalkBack on and Philoi off.
   *
   * Also the reason the feature degrades cleanly when the FOCUS_NUDGE_ANDROID build flag is off:
   * the config plugin then never registers the <service>, so this component can never appear in
   * that list, arm() returns false, and every layer above reads it as "not available on this
   * device" rather than as a failure.
   */
  private fun hasAccessibility(context: Context): Boolean {
    val expected = ComponentName(context, PhiloiFocusNudgeAccessibilityService::class.java)
    val enabled = Settings.Secure.getString(
      context.contentResolver,
      Settings.Secure.ENABLED_ACCESSIBILITY_SERVICES,
    ) ?: return false
    val splitter = TextUtils.SimpleStringSplitter(':')
    splitter.setString(enabled)
    while (splitter.hasNext()) {
      if (ComponentName.unflattenFromString(splitter.next()) == expected) return true
    }
    return false
  }

  private fun launchSettings(intent: Intent) {
    val activity = appContext.currentActivity
    if (activity != null) {
      activity.startActivity(intent)
      return
    }
    // No activity attached (the JS side raced a backgrounding). NEW_TASK is required to start an
    // activity from an application context and would leak the Settings screen into our own task if
    // used while one exists, hence the branch.
    context.startActivity(intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
  }
}
